Programmatically generate SSL key and certificate

ataylorme opened the pull request:
## Description
* Adds the `npm run generate-ssl-cert` command to generate an SSL key and certificate
* Removes the SSL certificate key and path options from `dev/config/themeConfig.js`
* Updates the `README` with instructions on generating an SSL key and certificate with WP Rig
* Updates the `README` with instructions on using your own generated SSL key and certificate with WP Rig

Addresses issue #81

## List of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] This pull request relates to a ticket.
- [x] My code is tested.
- [x] I want my code added to WP Rig.

ataylorme commented:
Note you still need to trust the generated key and certificate but they are good for 5 years and this should simplify enabling HTTPS for WP Rig.

@miklb it would be great if you could help test this.

ataylorme commented:
> Breaking change (fix or feature that would cause existing functionality to not work as expected)

If someone used the previous config to specify their own path for a key and certificate they will need to move them to the WP Rig specific paths listed in the README. I'm not sure if this is a breaking change as those options have only been in this `v1.1` branch and not officially released.

mor10 commented:
Not a breaking change since 1.1 is still pre-release. I will test (and probably merge) tomorrow.

ataylorme commented:
> that should read my own existing key, vs having to move/rename it to work with browser sync in WP Rig.

@miklb that is a good point. Folks can generate a key/cert once and reuse it across projects. I added the option to supply a custom key/cert path added back in.

I also updated the `README` to encourage a global key/cert across multiple projects.

ataylorme commented:
I've tested this thoroughly but would like at least one more person to test.

@miklb do you mind testing with your existing key/cert as well as generating them with WP Rig?

miklb commented:
I will be able to test later today

mor10 commented:
@miklb have you had a chance to test it yet?

miklb commented:
Apologies, I have not since it was updated. I will have time this evening.

ataylorme commented:
> Now my local dev environment is nginx running locally via homebrew and this was how I set up using .dev locally ssl.indieweb.org (before the Chrome issue with using .dev came about, but still works)

This shouldn't matter as BrowserSync is proxying that to `localhost:8181` so `localhost` is the domain as far as the browser is concerned.

> using the defaults for https: true I still get a insecure warning in Firefox.

I think the browsers have gotten stricter since I tested. Chrome now marks my (previously working) localhost proxy site as insecure as does Firefox.

They both seem to dislike that the certificate is self-signed and not sigend by a known certificate authority.

I'm not sure of a way to get around this. Any thoughts on generating a key and certificate that are properly signed are welcome.

miklb commented:
> This shouldn't matter as BrowserSync is proxying that to localhost:8181 so localhost is the domain as far as the browser is concerned.

Actually I think what I did was set it up so all traffic redirects to *.dev and I just add an entry in my `nginx.conf` file and the dev site is up. 

> I'm not sure of a way to get around this. Any thoughts on generating a key and certificate that are properly signed are welcome.

I'm going to keep thinking about it because I think it is important for devs to use local `https` and I'm excited this project is attempting to tackle that.

ataylorme commented:
Perhaps @Dexus would be kind enough to chime in and give us some help with pem :-)

miklb commented:
Looks like this might be what is needed https://www.npmjs.com/package/create-cert

it creates the ca cert that you can add to the browser and otherwise uses pem.

ataylorme commented:
That looks interesting but after tinkering/reading I don't see a way to load the CA cert into the proxy server BrowserSync creates.

miklb commented:
That ca cert gets added to the browser or keychain on Mac for Safari. That allows the browser to trust the key & cert

ataylorme commented:
Gotcha - should the CA cert file be saved as `wp-rig-browser-sync.pem` then?

miklb commented:
So you would run the gulp task to generate keys, import the ca cert then run main gulp task

ataylorme commented:
The `create-cert` package returns `key`, `cert` and `caCert`. I have to use Node's file system to save them. I know that `key` is `.key`, `cert` is `.crt` but I am not sure what extension I need to save the CA cert as.

miklb commented:
the CA cert is also `.cert` just need to make sure the two are named appropriately that we can document. I already have screenshots from how to add to Firefox for instance.

ataylorme commented:
I updated to use `create-cert` and added both the CA cert and regular cert to my KeyChain but get _This certificate has not been verified by a third-party_ in Chrome and _The certificate is not trusted because it is self-signed_ in FireFox.

@miklb want me to commit what I have? Maybe you can get it going. I am not an SSL expert and am beyond my expertise with keys/certs/CA signing.

miklb commented:
@ataylorme sounds good to me.

ataylorme commented:
@miklb you can see the changes in [`31bfdab3cfb8ab506ce06654936d567ea39fbcfb`](https://github.com/wprig/wprig/pull/92/commits/31bfdab3cfb8ab506ce06654936d567ea39fbcfb). 

`gulp/constants.js` defines the paths for the certs. Those paths are used when generating certs and by BrowserSync. The cert generation occurs in `gulp/generateCert.js`. Let me know if you have any questions.

miklb commented:
I was able to add the root cert to Firefox Dev Edition on macOS and aside from my localhost issue I think this works.

![wprig_ssl_add_root](https://user-images.githubusercontent.com/819716/45249511-75517280-b2ef-11e8-85a4-e40655d64981.png)

miklb commented:
@ataylorme are you sure you didn't try to add `wp-rig-browser-sync-cert.crt` to the browsers instead of the root cert? I believe that message about self-signed appears if you do.

mor10 commented:
This works for me. Are we good for a merge? Anyone have opinions?

bamadesigner commented:
@ataylorme Im testing now. Can you fix the merge conflict?

ataylorme commented:
> Can you fix the merge conflict?
Done

> Im testing now
@bamadesigner can you update the README (and maybe take screenshots) as you test? I think instructions on adding the certs to browsers is what is missing.

bamadesigner commented on package.json:
```diff
@@ -22,7 +22,8 @@
     "babel-register": "^6.26.0",
-    "browser-sync": "^2.23.6",
+    "browser-sync": "^2.24.7",
+    "create-cert": "^1.0.6",
```
I think we're missing `"generate-ssl-cert": "gulp generateCert",` under the scripts section.

And maybe we can change it to just be `generateCert` so: `"generateCert": "gulp generateCert",`

bamadesigner replied:
So then the instructions have to be changed to run `npm run generateCert`.

ataylorme replied:
Good catch, I missed the `genrateCert` command being lost in resolving the merge conflict. I added it back with the new name `generateCert` and updated the README.

bamadesigner replied:
Thanks!

bamadesigner commented on README.md:
```diff
@@ -47,21 +47,28 @@ WP Rig can be used in any development environment. It does not require any speci
         1. Open Keychain Access
-        2. Drag the certificate file onto KeyChain access
-        3. Double click on the new `WP Rig` certificate in the list
+        2. Drag the certificate files, which if using `npm run generateCert` to generate the certificates are saved to `BrowserSync/wp-rig-browser-sync-cert.crt` and `BrowserSync/wp-rig-browser-sync-root-cert.crt`, onto KeyChain access
```
I didn't need to trust `BrowserSync/wp-rig-browser-sync-root-cert.crt`.

I made a couple screenshots that I'll place into this section.

bamadesigner replied:
Also, when I dragged and dropped into Keychain Access, it just labelled it "localhost".

ataylorme replied:
Yeah, I don't think Keychain Access uses filename or anything just the hostname, which is `localhost`. That might be confusing if folks already have other `localhost` certificated but I don't think we can do anything about it.

bamadesigner commented:
@ataylorme let me know if you're ok with https://github.com/wprig/wprig/pull/92/commits/2d18f63d3a7caa1914298dabb674ea7055f9c870

ataylorme commented:
👍 looks good to me @bamadesigner. Were you able to get the green lock without warnings in Chrome and Firefox? I haven't been able to do so with only trusting the certs in Keychain Access and not the browser.

bamadesigner commented:
@ataylorme I was able to do so in Chrome and Safari. Not Firefox.

bamadesigner commented:
@ataylorme I'm going to merge and then I'll go in and add a few screenshots. It's easier in their editor.

miklb commented:
I can provide steps for adding to Firefox, I tested with FF and it was good to go.

bamadesigner commented:
@miklb thanks! If you share them here, I can add them to the readme.

miklb commented:
As soon as I’m back at the computer I’ll upload screenshots & paths to the settings

miklb commented:
Apologies for not posting these sooner, I had mentioned offering help with the docs and then got sidetracked. 
![firefox_add_cert_1](https://user-images.githubusercontent.com/819716/46054182-35262880-c114-11e8-9d0b-6b1157c6222d.png)
![firefox_add_cert_2](https://user-images.githubusercontent.com/819716/46054183-37888280-c114-11e8-961d-36e295d3b392.png)
![firefox_add_cert_wprig_3](https://user-images.githubusercontent.com/819716/46054743-9d760980-c116-11e8-9317-1bb539bdc26a.png)


I didn't add alt tags but text version - 

navigate to `about:preferences` in Firefox, scroll down to bottom of page under Certificates -> View Certificates

First, make sure the `Authorize` tab is active in the top nav bar. Second, click the middle bottom row Import -> navigate to directory of the of the certificates (wprig defaults to `/BrowserSync` select the root certificate `wp-rig-browser-sync-root-cert.crt` and import. If you are unable to select the certificate file, click the `options` tab to make sure `Certificate` file type is selected. (*Note* this last step doesn't seem to be an issue in Firefox Dev Edition I'm currently running, but was necessary a short time back in Firefox Nightly.)

bamadesigner commented:
Thanks @miklb! I hate to say this after you did the work but this seems like too much/too granular for including in our readme. Thoughts? @ataylorme? I wonder if there are some instructions we can link to instead?

ataylorme commented:
We could move all the HTTPS instructors to a wiki page here on GitHub and link to it from the README. That would condense the README and allow us to go into more depth explaination with screenshots.

My only concern is maintaining such a page as the instructions will surely change as browsers update. We’re already seeing discrepancies in the dev edition vs nightly of Firefox.

miklb commented:
Wiki page sounds good & a condensed version in readme sounds fine. I just wanted to be thorough. I documented those screenshots from own experience with my own local cert & had failed to share previously.

mor10 commented:
I like the Wiki page idea. Readme is starting to get a tad overloaded

miklb commented:
I’m really glad a solution was found & will ship in a future version. Thank you @ataylorme for all of the work on this.

bamadesigner commented:
I added all of the BrowserSync info to a wiki page: https://github.com/wprig/wprig/wiki/BrowserSync

bamadesigner commented:
I added all of the BrowserSync info to a wiki page: https://github.com/wprig/wprig/wiki/BrowserSync

And updated the readme to link, instead, to the wiki. I made note in the wiki that the generate certificate feature is still under development.

I'm going to merge. Let me know if anything is amiss.

Thank you for all of the hard work!

ataylorme commented:
Thank you @miklb and @bamadesigner for the documentation and testing - this was a team effort!
